src/next_array.rs

```diff
@@ -0,0 +1,64 @@
+use core::mem::MaybeUninit;
+use core::ptr;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N], // Invariant: arr[..len] is initialized.
+    len: usize,               // Invariant: len <= N.
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    pub fn new() -> Self {
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    pub fn push(&mut self, value: T) {
+        // We maintain the invariant here that arr[..len] is initialized.
+        // Indexing with self.len also ensures self.len < N, and thus <= N after
+        // the increment.
+        self.arr[self.len] = MaybeUninit::new(value);
+        self.len += 1;
+    }
```

Comment:
I really feel this is overly verbose and ultimately reduces readability of the code. In my opinion, sometimes less is more.

Reply:
I share that opinion, except in the case of documenting unsafe code. This PR is the first substantial addition of unsafe code to itertools, and I'd like us to set our standard for unsafe code very high. I also suspect that `ArrayBuilder` may turn out to be a major internal tool as we expand our support for const-generic adaptors.

Reply:
@jswrenn Of course I think the standard should be very high. And for that reason I think the comments should be pristine, and not overly verbose blocks of text that just make it harder to review the code as correct. I sincerely mean here that less is more, the shorter unsafe comments are *better* in my opinion.

Reply:
Think about it like this, lines of code that do things are signal, comments that do not help you understand the code are noise. In this example in your suggestion there are two lines of code, and nine lines of comments.

If you as a reader can reach the same understanding of the code with the same effort from a three-line comment as a nine-line comment (which I believe you can), then that means six lines were unnecessary and thus noise, giving a signal-to-noise ratio of 25%. I really prefer the signal-to-noise ratio to be as high as possible in `unsafe` code.

Reply:
We'll have to agree to disagree; you are unlikely to convince me otherwise.

Making these comments has been an essential step in vetting this PR, and these comments will continue to aid future contributors and reviewers in making and evaluating changes to this code. 

For example, I only noticed the (temporary) [invariant violation issue in `ArrayBuilder::take`](https://github.com/rust-itertools/itertools/pull/560#discussion_r1627861703) because I attempted to write detailed safety comments for this PR. I also see that you've accepted my suggested reordering, but not my suggested comments. Now, the effect achieved by reordering is totally implicit. We could easily regress on that issue.